python/ray/tune/sync_client.py

```diff
@@ -418,3 +433,235 @@ def _validate_exclude_template(exclude_template):
                     "Neither `{pattern}` nor `{regex_pattern}` found in "
                     f"exclude string `{exclude_template}`"
                 )
+
+
+def _get_recursive_files_and_stats(path: str) -> Dict[str, Tuple[float, int]]:
+    """Return dict of files mapping to stats in ``path``.
+
+    This function scans a directory ``path`` and returns a dict mapping
+    each contained file to a tuple of (mtime, filesize).
+    """
+    files_stats = {}
+    for root, dirs, files in os.walk(path, topdown=False):
+        rel_root = os.path.relpath(root, path)
+        for file in files:
+            key = os.path.join(rel_root, file)
+            stat = os.lstat(os.path.join(path, key))
+            files_stats[key] = round(stat.st_mtime, 5), stat.st_size
+
+    return files_stats
+
+
+# Only export once
+_remote_get_recursive_files_and_stats = ray.remote(_get_recursive_files_and_stats)
+
+
+@ray.remote
+def _pack_dir(
+    source_dir: str, files_stats: Optional[Dict[str, Tuple[float, int]]]
+) -> bytes:
+    stream = io.BytesIO()
+    with tarfile.open(fileobj=stream, mode="w:gz", format=tarfile.PAX_FORMAT) as tar:
+        if not files_stats:
+            # If no `files_stats` is passed, pack whole directory
+            tar.add(source_dir, arcname="", recursive=True)
+        else:
+            # Otherwise, only pack differing files
+            tar.add(source_dir, arcname="", recursive=False)
+            local_files_stats = _get_recursive_files_and_stats(source_dir)
+            for root, dirs, files in os.walk(source_dir, topdown=False):
+                rel_root = os.path.relpath(root, source_dir)
+                # Always add all directories
+                for dir in dirs:
+                    key = os.path.join(rel_root, dir)
+                    tar.add(os.path.join(source_dir, key), arcname=key, recursive=False)
+                # Add files where our information differs
+                for file in files:
+                    key = os.path.join(rel_root, file)
+                    if (
+                        key not in files_stats
+                        or local_files_stats[key] != files_stats[key]
+                    ):
+                        tar.add(os.path.join(source_dir, key), arcname=key)
+
+    return stream.getvalue()
+
+
+@ray.remote
+def _unpack_dir(stream: bytes, target_dir: str):
+    with tarfile.open(fileobj=io.BytesIO(stream)) as tar:
+        tar.extractall(target_dir)
+
+
+@ray.remote
+def _delete_dir(target_dir: str) -> bool:
+    if os.path.exists(target_dir):
+        shutil.rmtree(target_dir)
+        return True
+    return False
+
+
+class RemoteTaskClient(SyncClient):
```

Comment:
nit nit, RemoteTaskSyncClient?

Reply:
The other classes are called `CommandBasedClient` and `FunctionBasedClient`, so `RemoteTaskClient` is actually consistent

Reply:
oh 👌